File: python/pascals-triangle/pascals_triangle.py
def rows(row_count):
    if row_count < 0:
        raise ValueError("number of rows is negative")

    return row_recurse([[1]], row_count) if row_count > 0 else []


def row_recurse(so_far, row_count):
    prev = so_far[-1]
    if len(prev) == row_count:
        return so_far
    so_far.append(make_row([1], prev))
    return row_recurse(so_far, row_count)


def make_row(accum, rest):
    if rest == [1]:
        accum.append(1)
        return accum
    else:
        accum.append(rest[0] + rest[1])
        return make_row(accum, rest[1:])

File: julia/pascals-triangle/pascals-triangle.jl
function row_recurse(so_far, n)
    prev = so_far[end]
    length(prev) == n && return so_far
    push!(so_far, make_row([1], prev))
    row_recurse(so_far, n)
end

function make_row(accum, rest)
    rest == [1] && return vcat(accum, 1)
    push!(accum, rest[1] + rest[2])
    make_row(accum, rest[2:end])
end

function triangle(n)
    n < 0 && throw(DomainError(n, "n must be non-negative"))
    n > 0 ? row_recurse([[1]], n) : []
end
